File: Go/Euler020.go
package main

import (
	"fmt"
	"math/big"
)

func main() {
	const targetInt = 100
	var target *big.Int
	target = big.NewInt(targetInt)

	result := factorial(target)

	big10 := big.NewInt(10)
	big0 := big.NewInt(0)

	var sum int64
	digit := big.NewInt(0)

	for result.Cmp(big0) == 1 {
		result.DivMod(result, big10, digit)
		sum += digit.Int64()
	}

	fmt.Println(sum)
}

func factorial(n *big.Int) *big.Int {
	big1 := big.NewInt(1)
	fact := big.NewInt(1)

	for counter := n; n.Cmp(big1) == 1; counter.Sub(counter, big1) {
		fact.Mul(fact, n)
	}
	return fact
}

File: Kotlin/Euler020.kt
package Euler020

import java.math.BigInteger

fun main() {
  val target = 100
  val big0 = 0.toBigInteger()
  val big10 = 10.toBigInteger()

  var result = factorial(target)
  var sum = 0.toBigInteger()

  while (result > big0) {
    sum += result.mod(big10)
    result = result.div(big10)
  }

  println(sum)
}

fun factorial(n: Int): BigInteger {
  val product = (n downTo 1).asSequence().map { x -> x.toBigInteger() }.reduce { a, b -> a.multiply(b) }
  return product
}
